File: src/test/java/org/embulk/input/TestRemoteFileInputPlugin.java
package org.embulk.input;

import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import com.github.dockerjava.api.DockerClient;
import com.github.dockerjava.api.model.Container;
import com.github.dockerjava.core.DockerClientBuilder;
import org.embulk.EmbulkEmbed;
import org.embulk.config.ConfigSource;
import org.embulk.spi.InputPlugin;
import org.embulk.test.MemoryOutputPlugin;
import org.embulk.test.MyEmbulkTests;
import org.embulk.test.MyTestingEmbulk;
import org.embulk.test.TestingEmbulk;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

public class TestRemoteFileInputPlugin {
    private static final String CONTAINER_ID_HOST1 = "embulkinputremote_host1_1";
    private static final String CONTAINER_ID_HOST2 = "embulkinputremote_host2_1";
    private static final DockerClient dockerClient = DockerClientBuilder.getInstance().build();

    @Rule
    public MyTestingEmbulk embulk = (MyTestingEmbulk) MyTestingEmbulk
            .builder()
            .registerPlugin(InputPlugin.class, "remote", RemoteFileInputPlugin.class)
            .build();

    @Before
    public void prepare() {
        startContainer(CONTAINER_ID_HOST1);
        startContainer(CONTAINER_ID_HOST2);

        String logLevel = System.getenv("LOG_LEVEL");
        if (logLevel != null) {
            // Set log level
            Logger rootLogger = (Logger) LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME);
            rootLogger.setLevel(Level.toLevel(logLevel));
        }
    }

    @Test
    public void loadFromRemote() throws Exception
    {
        embulk.runInput(baseConfig());
        assertValues(values(1L, "user1"));
    }

    @Ignore("Cannot pass on TravisCI, although pass on Local Mac OS...")
    @Test
    public void loadFromRemoteViaPublicKey() throws Exception
    {
        String keyPath = System.getenv("KEY_PATH");
        if (keyPath == null) {
            keyPath = "./id_rsa_test";
        }

        final ConfigSource publicKeyAuth = newConfig().set("auth", newConfig()
                .set("type", "public_key")
                .set("key_path", keyPath)
        );
        embulk.runInput(baseConfig().merge(publicKeyAuth));

        assertValues(values(1L, "user1"));
    }

    @Test
    public void testMultiHosts() throws Exception
    {
        final ConfigSource multiHosts = newConfig()
                .set("hosts", Arrays.asList("localhost:10022", "localhost:10023"));
        final ConfigSource config = baseConfig().merge(multiHosts);

        // Run
        embulk.runInput(config);
        assertValues(
                values(1L, "user1"),
                values(2L, "user2")
        );
    }

    @Test
    public void loadAllFilesInDirectory() throws Exception
    {
        final ConfigSource multiHosts = newConfig()
                .set("path", "/mount");
        final ConfigSource config = baseConfig().merge(multiHosts);

        embulk.runInput(config);
        assertValues(
                values(1L, "user1"),
                values(1L, "command_user1")
        );
    }

    @Test
    public void testDefaultPort() throws Exception
    {
        final ConfigSource defaultPort = newConfig()
                .set("hosts", Collections.singletonList("localhost"))
                .set("default_port", 10022);

        embulk.runInput(baseConfig().merge(defaultPort));

        assertValues(values(1L, "user1"));
    }

    @Test
    public void testConfDiff() throws Exception
    {
        final ConfigSource host2Config = newConfig()
                .set("hosts", Collections.singletonList("localhost:10023"));
        ConfigSource config = baseConfig().merge(host2Config);

        // Run
        TestingEmbulk.RunResult runResult = embulk.runInput(config);
        assertValues(values(2L, "user2"));

        // Re-run with additional host1
        final ConfigSource multiHost = newConfig()
                .set("hosts", Arrays.asList("localhost:10022", "localhost:10023"));
        config = baseConfig().merge(multiHost);

        embulk.runInput(config, runResult.getConfigDiff());

        assertValues(values(1L, "user1"));
    }

    @Test
    public void testResume() throws Exception
    {
        final ConfigSource multiHost = newConfig()
                .set("hosts", Arrays.asList("localhost:10022", "localhost:10023"));
        final ConfigSource config = baseConfig().merge(multiHost);

        // Stop host2 temporarily
        stopContainer(CONTAINER_ID_HOST2);

        // Run (but will fail)
        EmbulkEmbed.ResumableResult resumableResult = embulk.resume(config);

        assertThat(resumableResult.isSuccessful(), is(false));
        assertValues(values(1L, "user1"));

        // Start host2 again
        startContainer(CONTAINER_ID_HOST2);

        // Resume
        resumableResult = embulk.resume(config, resumableResult.getResumeState());

        assertThat(resumableResult.isSuccessful(), is(true));
        assertValues(values(2L, "user2"));
    }

    @Test
    public void testIgnoreNotFoundHosts() throws Exception
    {
        final ConfigSource ignoreNotFoundHosts = newConfig()
                .set("hosts", Arrays.asList("localhost:10022", "localhost:10023"))
                .set("ignore_not_found_hosts", true);
        final ConfigSource config = baseConfig().merge(ignoreNotFoundHosts);

        // Stop host2
        stopContainer(CONTAINER_ID_HOST2);

        // Run (host2 will be ignored)
        EmbulkEmbed.ResumableResult resumableResult = embulk.resume(config);

        assertThat(resumableResult.isSuccessful(), is(true));
        assertValues(values(1L, "user1"));
    }

    @Test
    public void testCommandOptions() throws Exception
    {
        final ConfigSource ignoreNotFoundHosts = newConfig()
                .set("hosts_command", "echo localhost:10022,localhost:10023")
                .set("hosts_separator", ",")
                .set("path_command", "echo /mount/test_command.csv");
        final ConfigSource config = baseConfig().merge(ignoreNotFoundHosts);

        embulk.runInput(config);

        assertValues(
                values(1L, "command_user1"),
                values(2L, "command_user2")
        );
    }

    //////////////////////////////
    // Helpers
    //////////////////////////////

    private ConfigSource baseConfig() {
        return MyEmbulkTests.configFromResource("yaml/base.yml");
    }

    private ConfigSource newConfig() {
        return embulk.newConfig();
    }

    private void assertValues(List... valuesList) {
        Set<List> actual = new HashSet<>();
        for (MemoryOutputPlugin.Record record : MemoryOutputPlugin.getRecords()) {
            actual.add(record.getValues());
        }

        Set<List> expected = new HashSet<>();
        Collections.addAll(expected, valuesList);

        assertThat(actual, is(expected));
    }

    private List values(Object... values) {
        return Arrays.asList(values);
    }

    //////////////////////////////
    // Methods for Docker
    //////////////////////////////

    private static void stopContainer(String containerId) {
        if (isRunning(containerId)) {
            dockerClient.stopContainerCmd(containerId).exec();
        }
    }

    private static void startContainer(String containerId) {
        if (!isRunning(containerId)) {
            dockerClient.startContainerCmd(containerId).exec();
        }
    }

    private static boolean isRunning(String containerId) {
        List<Container> containers = dockerClient.listContainersCmd().exec();
        for (Container container : containers) {
            for (String name : container.getNames()) {
                if (name.contains(containerId)) {
                    System.out.println("Found " + containerId);
                    return true;
                }
            }
        }
        System.out.println("Not Found " + containerId);
        return false;
    }
}

File: src/test/kotlin/org/embulk/input/TestRemoteFileInputPlugin.kt
package org.embulk.input

import ch.qos.logback.classic.Level
import ch.qos.logback.classic.Logger
import com.github.dockerjava.core.DockerClientBuilder
import org.embulk.config.ConfigSource
import org.embulk.test.EmbulkPluginTest
import org.embulk.test.TestOutputPlugin.Matcher.assertRecords
import org.embulk.test.configFromResource
import org.embulk.test.record
import org.embulk.test.registerPlugin
import org.embulk.test.set
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before
import org.junit.Ignore
import org.junit.Test
import org.slf4j.LoggerFactory

class TestRemoteFileInputPlugin : EmbulkPluginTest() {

    @Before fun setup() {
        builder.registerPlugin(RemoteFileInputPlugin::class, "remote")

        // Setup docker container
        startContainer(CONTAINER_ID_HOST1)
        startContainer(CONTAINER_ID_HOST2)

        System.getenv("LOG_LEVEL")?.let {
            // Set log level
            val rootLogger = LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME) as Logger
            rootLogger.level = Level.toLevel(it)
        }
    }

    @Test fun loadFromRemote() {
        runInput(baseConfig())
        assertRecords(record(1, "user1"))
    }

    @Ignore("Cannot pass on TravisCI, although pass on Local Mac OS...")
    @Test fun loadFromRemoteViaPublicKey() {
        val keyPath = System.getenv("KEY_PATH") ?: "./id_rsa_test"

        val publicKeyAuth = config().set("auth" to config().set(
                "type" to "public_key",
                "key_path" to keyPath
        ))
        runInput(baseConfig().merge(publicKeyAuth))

        assertRecords(record(1, "user1"))
    }

    @Test fun testMultiHosts() {
        val multiHosts = config()
                .set("hosts", listOf("localhost:10022", "localhost:10023"))

        // Run
        runInput(baseConfig().merge(multiHosts))
        assertRecords(
                record(1, "user1"),
                record(2, "user2")
        )
    }

    @Test fun loadAllFilesInDirectory() {
        val directoryPath = config().set("path", "/mount")

        runInput(baseConfig().merge(directoryPath))
        assertRecords(
                record(1L, "user1"),
                record(1L, "command_user1")
        )
    }

    @Test fun testDefaultPort() {
        val defaultPort = config().set(
                "hosts" to listOf("localhost"),
                "default_port" to 10022
        )

        runInput(baseConfig().merge(defaultPort))

        assertRecords(record(1L, "user1"))
    }

    @Test fun testConfDiff() {
        val host2Config = config().set("hosts", listOf("localhost:10023"))

        // Run
        val runResult = runInput(baseConfig().merge(host2Config))
        assertRecords(record(2, "user2"))

        // Re-run with additional host1
        val multiHost = config().set("hosts", listOf("localhost:10022", "localhost:10023"))
        runInput(baseConfig().merge(multiHost), runResult.configDiff)

        assertRecords(record(1, "user1"))
    }

    @Test fun testResume() {
        // Stop host2 temporarily
        stopContainer(CONTAINER_ID_HOST2)

        // Run (but will fail)
        val multiHost = config().set("hosts", listOf("localhost:10022", "localhost:10023"))
        val config = baseConfig().merge(multiHost)
        var resumableResult = resume(config)

        assertThat(resumableResult.isSuccessful, `is`(false))
        assertRecords(record(1, "user1"))

        // Start host2 again
        startContainer(CONTAINER_ID_HOST2)

        // Resume
        resumableResult = resume(config, resumableResult.resumeState)

        assertThat(resumableResult.isSuccessful, `is`(true))
        assertRecords(record(2, "user2"))
    }

    @Test fun testIgnoreNotFoundHosts() {
        val ignoreNotFoundHosts = config().set(
                "hosts" to listOf("localhost:10022", "localhost:10023"),
                "ignore_not_found_hosts" to true
        )
        val config = baseConfig().merge(ignoreNotFoundHosts)

        // Stop host2
        stopContainer(CONTAINER_ID_HOST2)

        // Run (host2 will be ignored)
        val resumableResult = resume(config)

        assertThat<Boolean>(resumableResult.isSuccessful, `is`(true))
        assertRecords(record(1, "user1"))
    }

    @Test fun testCommandOptions() {
        val ignoreNotFoundHosts = config().set(
                "hosts_command" to "./src/test/resources/script/hosts.sh",
                "hosts_separator" to "\n",
                "path_command" to "echo '/mount/test_command.csv'"
        )
        runInput(baseConfig().merge(ignoreNotFoundHosts))

        assertRecords(
                record(1, "command_user1"),
                record(2, "command_user2")
        )
    }

    //////////////////////////////
    // Helpers
    //////////////////////////////

    private fun baseConfig(): ConfigSource {
        return configFromResource("yaml/base.yml")
    }

    companion object DockerUtils {
        private val CONTAINER_ID_HOST1 = "embulkinputremote_host1_1"
        private val CONTAINER_ID_HOST2 = "embulkinputremote_host2_1"
        private val dockerClient = DockerClientBuilder.getInstance().build()

        private fun stopContainer(containerId: String) {
            if (isRunning(containerId)) {
                dockerClient.stopContainerCmd(containerId).exec()
            }
        }

        private fun startContainer(containerId: String) {
            if (!isRunning(containerId)) {
                dockerClient.startContainerCmd(containerId).exec()
            }
        }

        private fun isRunning(containerId: String): Boolean {
            return dockerClient.listContainersCmd().exec().any { container ->
                container.names.any { name ->
                    name.contains(containerId)
                }
            }
        }
    }
}
